### PR TITLE
690 nan inf loss

### LINKED ISSUE #690: Nan/inf loss encountered
## Subject of the issue

More discussion is here https://github.com/Project-MONAI/MONAI/issues/1868

Using script https://github.com/DeepRegNet/Benchmark/blob/9-data-for-miccai/benchmark/balakrishnan2019/voxel_morph_balakrishnan_2019.py with the latest `main` branch.

```
226/226 [==============================] - 3083s 14s/step - loss: -0.7058 - loss/regularization_GradientNorm: 2.1987e-04 - loss/regularization_GradientNorm_weighted: 2.1987e-04 - loss/image_LocalNormalizedCrossCorrelationLoss: -0.7060 - loss/image_LocalNormalizedCrossCorrelationLoss_weighted: -0.7060
Epoch 2/2000
226/226 [==============================] - 3266s 14s/step - loss: -0.4778 - loss/regularization_GradientNorm: 0.3009 - loss/regularization_GradientNorm_weighted: 0.3009 - loss/image_LocalNormalizedCrossCorrelationLoss: -0.7787 - loss/image_LocalNormalizedCrossCorrelationLoss_weighted: -0.7787
Epoch 3/2000
226/226 [==============================] - 3280s 15s/step - loss: nan - loss/regularization_GradientNorm: nan - loss/regularization_GradientNorm_weighted: nan - loss/image_LocalNormalizedCrossCorrelationLoss: -inf - loss/image_LocalNormalizedCrossCorrelationLoss_weighted: -inf
```

## Your environment

- DeepReg version (commit hash)

  `5e5bbdc`

- OS (e.g. Ubuntu 20.04, MacOS 10.15, etc.)

  Ubuntu

- Python Version (3.7, 3.8, etc.)

  3.7

- TensorFlow

  2.3

### PR DESCRIPTION
**Pull request by mathpluscode**
# Description

After a long debugging, I found the final source of inf was due to a negative discrete variance (~1e-5) in LNCC loss function. This can happen is that, within a window, the voxels might have the same values thus the variance should be 0, then numerical issues causes (somehow) a negative value in the end.

The fix is to normalize the kernel weights directly so that we do not need to divide the kernel volume later. During debugging, the min variance is now ~-1e-11, althou still negative, it's around machine error now.

To be safe, we now clip the variance to 0 and add an EPS 1e-5 (EPS was 1e-7, but VoxelMorph adopts 1e-5 and if we consider mixed-precision later, 1e-5 will be safer).

Fixes #690

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [ ] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by YipengHu — APPROVED** [review]
@mathpluscode can you double-check if there is anywhere multiple separable kernels was used and normalised afterwards - which should be redundant now.

**Comment by mathpluscode** [comment]
I've been testing the new code overnight and no more inf/crash ^^

**Comment by YipengHu** [comment]
for record: 
[Project-MONAI/MONAI#1868](https://github.com/Project-MONAI/MONAI/issues/1868)

**Comment by YipengHu** [comment]
@mathpluscode have you tested if the 2-pass algorithm produces negative var?

**Comment by mathpluscode** [comment]
> @mathpluscode have you tested if the 2-pass algorithm produces negative var?

It won't as you do conv on the square of tensors, and conv having all positive weights. So by design, this will be >= 0.

**Comment by YipengHu** [comment]
> > @mathpluscode have you tested if the 2-pass algorithm produces negative var?
> 
> It won't as you do conv on the square of tensors, and conv having all positive weights. So by design, this will be >= 0.
> 
> I've tested the current implementation for multiple epochs and it seems to be safe.

so it won;t have the inf issue

**Comment by mathpluscode** [comment]
> > > @mathpluscode have you tested if the 2-pass algorithm produces negative var?
> > 
> > 
> > It won't as you do conv on the square of tensors, and conv having all positive weights. So by design, this will be >= 0.
> > I've tested the current implementation for multiple epochs and it seems to be safe.
> 
> so it won;t have the inf issue

It should not.

**Review by YipengHu — APPROVED** [review]

**Comment by YipengHu** [comment]
@acasamitjana can you have a look at this please?

**Review by acasamitjana — APPROVED** [review]
I've seen some issues with negative variance in tensorflow. Good catch and nice solution @mathpluscode
